library/cipher.c

```diff
@@ -1468,6 +1479,54 @@ static int mbedtls_cipher_aead_decrypt( mbedtls_cipher_context_t *ctx,
 
     return( MBEDTLS_ERR_CIPHER_FEATURE_UNAVAILABLE );
 }
+
+#if !defined(MBEDTLS_DEPRECATED_REMOVED)
```

Comment:
As noted above, re-introducing deprecated functions is probably a mishap in previous conflict resolution.

Reply:
> I have reviewed the commits up to [c40bc1e](https://github.com/ARMmbed/mbedtls/commit/c40bc1e406eb67b31752a9586e49b79e4211ce48) “Fix Changelog typo”, except for some test functions (noted in a comment) and for the test data (which I have neither reviewed for correctness nor completeness, and I've requested quite a bit more coverage).
> 
I merged the work done to add proper multipart into this to try and avoid a force push, as at that point (see previous comments) it would not compile. I did not merge all of the work as this caused an amount of conflicts that would not have been easy to resolve and I merged the files as they were when they were pushed to try and avoid any other rebase issues, however this has resulted in a load of re-introduced issues that are already fixed in development. 

Its fairly clear then, that this whole thing is going to require another rebase to put it up to development now (the origin of this branch is over a month old at this point, and suffers from issues with generated files, as well as the gcm api changes, and a number of commits could then be removed).

My question would be when to do this, as a force push will obviously cause review issues. I could do this after I have dealt with the current outstanding issues, would that be acceptable? The 2.x branch suffers from none of these issues currently.

Reply:
As far as I'm concerned, it's ok to rebase now. However, given the large number of commits and review comments, I suggest an actual merge. The merge commit should have detailed comments explaining conflict resolution. See `git log --merges --author=Gilles.Peskine --grep=Conflict` for some examples of what I consider to be adequate documentation of merge conflict resolution. The merge commit should minimize changes that are not conflict resolution (and if there are any, document them precisely). Other adaptations, for example to get tests to pass, should be in subsequent commits.